atom/browser/atom_download_manager_delegate.cc

```diff
@@ -11,23 +11,294 @@
 #include "atom/browser/ui/file_dialog.h"
 #include "base/bind.h"
 #include "base/files/file_util.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/download/download_completion_blocker.h"
+#include "chrome/browser/download/download_item_model.h"
 #include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/safe_browsing/download_protection/download_protection_util.h"
+#include "chrome/browser/safe_browsing/safe_browsing_service.h"
 #include "chrome/common/pref_names.h"
+#include "chrome/common/safe_browsing/file_type_policies.h"
+#include "chrome/grit/generated_resources.h"
 #include "components/prefs/pref_service.h"
 #include "content/public/browser/browser_context.h"
 #include "content/public/browser/browser_thread.h"
+#include "content/public/browser/download_danger_type.h"
 #include "content/public/browser/download_manager.h"
 #include "net/base/filename_util.h"
+#include "ui/base/l10n/l10n_util.h"
 
 namespace atom {
 
 namespace {
 
+using content::BrowserThread;
+using content::DownloadItem;
+using content::DownloadManager;
+using safe_browsing::DownloadFileType;
+using safe_browsing::DownloadProtectionService;
+
+#if defined(FULL_SAFE_BROWSING)
+
+// String pointer used for identifying safebrowing data associated with
+// a download item.
+const char kSafeBrowsingUserDataKey[] = "Safe Browsing ID";
+
+// The state of a safebrowsing check.
+class SafeBrowsingState : public DownloadCompletionBlocker {
+ public:
+  SafeBrowsingState() {}
+  ~SafeBrowsingState() override;
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(SafeBrowsingState);
+};
+
+SafeBrowsingState::~SafeBrowsingState() {}
+
+void CheckDownloadUrlDone(
```

Comment:
seems like we're copying a lot of code from src/chrome here which we really try to avoid. Can we subclass or use some other method to reuse the existing code?

Reply:
Most of the code is similar to `chrome_download_manager_delegate.cc`. I optimized some of it, let me know if you think we can optimize more.